client/src/store/utils/reducerFunctions.js

```diff
@@ -23,6 +23,60 @@ export const addMessageToStore = (state, payload) => {
   });
 };
 
+export const updateMessageReadInStore = (state, senderId, conversationId) => {
+  const conversation = state.filter(conversation=> conversation.id === conversationId)[0];
+  const messages = conversation.messages.map((message) => {
+    if (message.read === false && message.senderId !== senderId) {
+      const messageCopy = {...message}
+      messageCopy.read = true;
+      return messageCopy;
+    } else {
+      return message;
+    }
+  });
+
+  return state.map((conversation)=>{
+    if (conversation.id == conversationId){
+      const newConvo = {...conversation}
+      newConvo.messages = messages;
+      newConvo.messages.forEach(message=>{
+        if (message.id > newConvo.lastReadMessageId && message.read === true && message.senderId !== senderId)
+          {
+            newConvo.lastReadMessageId = message.id
+          }
+      });
+      newConvo.unreadMessagesCount = newConvo.messages.filter(message => message.read === false 
+        && message.senderId === newConvo.otherUser.id).length;
+      return newConvo;
+    } else {
+      return conversation
+    }
+    });
+}
+
+export const markStoreMessagesRead = (state, conversationId) => {
+  const conversation = state.filter(conversation=> conversation.id === conversationId)[0];
+  const messages = conversation.messages.map((message) => {
+    if (message.read === false) {
+      const messageCopy = {...message}
+      messageCopy.read = true;
+      return messageCopy;
+    } else {
+      return message;
+    }
+  });
+
+  return state.map((conversation)=>{
+    if (conversation.id == conversationId){
+      const newConvo = {...conversation}
+      newConvo.messages = messages;
+      return newConvo;
+    } else {
+      return conversation
+    }
+    });
```

Comment:
This logic could be simplified. Same as above, you loop twice to find conversations and to update messages.

Reply:
Same as above, the two loops are mutually exclusive. If you can provide an example of what you mean by simplifying it further, that could be helpful. Although i can't see at the moment any reason to use one loop